### PR TITLE
user self registration

### PR DESCRIPTION
**Pull request by mikems66**

### CODE REVIEW

File: modern/src/LoginPage.js [hunk 1]
```diff
@@ -118,6 +137,12 @@ const LoginPage = () => {
             </div>
           </FormControl>
         </form>
+        {registerDialogShown && <RegisterDialog open={registerDialogShown} onResult={handleRegisterResult} />}
+        <Snackbar
```
Comment (tananaev) [inline modern/src/LoginPage.js:141]: Please follow existing code formatting style.

Comment (tananaev) [reply modern/src/LoginPage.js:141]: Also, is it possible to move snackbar to the `RegisterDialog` component?

File: modern/src/RegisterDialog.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+          required
+          fullWidth
+          error={validationErrors.password}
+          label={t('userPassword')}
+          name='password'
+          value={formFields.password || ''}
+          type='password'
+          autoComplete='current-password'
+          onChange={handleChange}
+          helperText={validationErrors.password && 'Password is required'} />
```
Comment (tananaev) [inline modern/src/RegisterDialog.js:98]: Please don't use unlocalized text anywhere.

File: modern/src/RegisterDialog.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  const handleRegister = async (event) => {
+    event.preventDefault();
+    setErrorResponse(null);
+    let objErrors = {};
+    if (formFields.name.trim() === '') {
+      objErrors.name = true;
+    }
+    if (!
```
Comment (tananaev) [inline modern/src/RegisterDialog.js:28]: Why is there a new line here?

File: modern/src/RegisterDialog.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    const response = await fetch('/api/users', {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(formFields)
+    });
+
+    if (response.ok) {
+      onResult(true)
+    } else {
+      setErrorResponse(t('errorGeneral'));
```
Comment (tananaev) [inline modern/src/RegisterDialog.js:49]: I would skip this error check for now. We need more standardized error handing for all calls.

File: modern/src/RegisterDialog.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  const [validationErrors, setValidationErrors] = useState({});
+  const [errorResponse, setErrorResponse] = useState(null);
+
+  const handleChange = (event) => {
+    setFormFields({ ...formFields, [event.target.name]: event.target.value });
+    setValidationErrors({ ...validationErrors, [event.target.name]: false });
+  }
+
+  const handleRegister = async (event) => {
+    event.preventDefault();
```
Comment (tananaev) [inline modern/src/RegisterDialog.js:22]: Is this required?

File: modern/src/RegisterDialog.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+import Button from '@material-ui/core/Button';
+import Dialog from '@material-ui/core/Dialog';
+import DialogActions from '@material-ui/core/DialogActions';
+import DialogContent from '@material-ui/core/DialogContent';
+import DialogContentText from '@material-ui/core/DialogContentText';
+import TextField from '@material-ui/core/TextField';
+import Alert from '@material-ui/lab/Alert';
+
+const RegisterDialog = ({ open, onResult }) => {
+  const [formFields, setFormFields] = useState({});
```
Comment (tananaev) [inline modern/src/RegisterDialog.js:12]: Why not have separate fields?

File: modern/src/RegisterDialog.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  const [validationErrors, setValidationErrors] = useState({});
```
Comment (tananaev) [inline modern/src/RegisterDialog.js:13]: Same here.

File: modern/src/LoginPage.js [hunk 8]
```diff
@@ -47,10 +50,19 @@ const LoginPage = () => {
...
+  const [registerDialogShown, setRegisterDialogShown] = useState(false);
+  const [snackbarOpen, setSnackbarOpen] = useState(false);
 
   const classes = useStyles();
   const history = useHistory();
 
+  const server = useSelector(state => state.session.server);
+
+  let registrationPref = false;
+  if (server) {
```
[the remark this answers is not in the document]
Comment (mikems66) [reply modern/src/LoginPage.js]: Thanks for your feedback.
I did it in the useSelector.

File: modern/src/LoginPage.js (hunk 1 above)
Comment (mikems66) [reply modern/src/LoginPage.js]: I hope formatting is good now.
I moved snack bar to RegisterDialog.

File: modern/src/RegisterDialog.js (hunk 2 above)
Comment (mikems66) [reply modern/src/RegisterDialog.js]: I changed this to existing localized string. If better labels are needed, let me know.

File: modern/src/RegisterDialog.js (hunk 3 above)
Comment (mikems66) [reply modern/src/RegisterDialog.js]: removed line

File: modern/src/RegisterDialog.js (hunk 4 above)
Comment (mikems66) [reply modern/src/RegisterDialog.js]: Error checking is skipped now.

File: modern/src/RegisterDialog.js (hunk 5 above)
Comment (mikems66) [reply modern/src/RegisterDialog.js]: No, removed.

File: modern/src/RegisterDialog.js (hunk 6 above)
Comment (mikems66) [reply modern/src/RegisterDialog.js:12]: I didn't change, I think this way code is more compact.

File: modern/src/RegisterDialog.js (hunk 7 above)
Comment (mikems66) [reply modern/src/RegisterDialog.js:13]: Same as above.

File: modern/src/RegisterDialog.js [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+      setSnackbarOpen(true);
+    }
+  }
+
+  if (snackbarOpen) {
+    return <Snackbar
+      anchorOrigin={{ vertical: 'bottom', horizontal: 'center', }}
+      open={snackbarOpen} autoHideDuration={6000}
+      onClose={() => { onResult(true) }} message={t('loginCreated')}
+    />
```
Comment (tananaev) [inline modern/src/RegisterDialog.js:54]:
```suggestion
    return (
      <Snackbar
        anchorOrigin={{ vertical: 'bottom', horizontal: 'center' }}
        open={snackbarOpen}
        autoHideDuration={6000}
        onClose={() => { onResult(true) }}
        message={t('loginCreated')} />
    );
```

File: modern/src/RegisterDialog.js [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+          error={validationErrors.email}
+          label={t('userEmail')}
+          name='email'
+          value={formFields.email || ''}
+          autoComplete='email'
+          onChange={handleChange}
+          helperText={validationErrors.email && t('sharedRequired')} />
+        <TextField
+          margin='normal'
+          required
```
Comment (tananaev) [inline modern/src/RegisterDialog.js:86]: Doesn't thins already handle "required"? Why do we need extra validation?

Comment (mikems66) [reply modern/src/RegisterDialog.js]: It allowes all blanks to be entered as name and password. But maybe that's not important, so I removed validation (except e-mail).

Comment (tananaev) [reply modern/src/RegisterDialog.js]: There is also a standard email validation, as far as I know. Can we use that?

File: modern/src/RegisterDialog.js (hunk 6 above)
Comment (tananaev) [reply modern/src/RegisterDialog.js:12]: I think we should separate fields to match what we have in other places. And errors is not needed at all, I think.

Comment (mikems66) [reply modern/src/RegisterDialog.js:12]: I will separate fields.

File: modern/src/RegisterDialog.js (hunk 10 above)
Comment (mikems66) [reply modern/src/RegisterDialog.js]: If you mean use type='email', seems this doesn't work with Dialog component (submit still works). I could disable submit button like this:  
disabled={!userName || !/(.+)@(.+)\.(.{2,})/.test(email) || !password}  ... let me know if that's ok.

Comment (tananaev) [reply modern/src/RegisterDialog.js]: Yeah, let's do something like that, but maybe move that regex to a separate method for readability.

**Comment by tananaev** [comment]
Merged, thank you. I will done some minor updates to it on master.

**Comment by mikems66** [comment]
No problem. I could do more stuff, let me know if you have any suggestions.

**Comment by tananaev** [comment]
One of the biggest things that's not implemented yet is geofences, but that might be too complicated.

**Pull request merged**